csharp/src/Ice/InputStream.cs

```diff
@@ -592,456 +208,526 @@ public long ReadLong()
-                SkipSlice();
+        //
+        // Internal and private methods
+        //
 
-                if ((_current.SliceFlags & EncodingDefinitions.SliceFlags.IsLastSlice) != 0)
-                {
-                    // Create and throw a plain RemoteException with the SlicedData.
-                    Debug.Assert(SlicedData != null);
-                    remoteEx = new RemoteException(SlicedData.Value);
-                    remoteEx.ConvertToUnhandled = true;
-                    return remoteEx;
-                }
+        /// <summary>Reads an empty encapsulation from the provided byte buffer.</summary>
+        /// <param name="communicator">The communicator.</param>
+        /// <param name="buffer">The byte buffer.</param>
+        internal static void ReadEmptyEncapsulation(Communicator communicator, ArraySegment<byte> buffer)
+        {
+            var istr = new InputStream(communicator, buffer, startEncaps: true, 0);
+            istr.SkipTaggedParams();
+            istr.CheckEndOfBuffer();
+        }
 
-                typeId = ReadSliceHeaderIntoCurrent()!;
-                ReadIndirectionTableIntoCurrent();
-            }
+        /// <summary>Reads the contents of an encapsulation from the provided byte buffer.</summary>
+        /// <param name="communicator">The communicator.</param>
+        /// <param name="buffer">The byte buffer.</param>
+        /// <param name="payloadReader">The reader used to read the payload of this encapsulation.</param>
+        internal static T ReadEncapsulation<T>(Communicator communicator,
+                                               ArraySegment<byte> buffer,
+                                               InputStreamReader<T> payloadReader)
+        {
+            var istr = new InputStream(communicator, buffer, startEncaps: true, 0);
+            T result = payloadReader(istr);
+            istr.SkipTaggedParams();
+            istr.CheckEndOfBuffer();
+            return result;
         }
 
-        /// <summary>
-        /// Skip the given number of bytes.
-        /// </summary>
-        /// <param name="size">The number of bytes to skip</param>
-        public void Skip(int size)
+        internal static int ReadInt(ReadOnlySpan<byte> buffer) => BitConverter.ToInt32(buffer);
+        internal static long ReadLong(ReadOnlySpan<byte> buffer) => BitConverter.ToInt64(buffer);
+        internal static short ReadShort(ReadOnlySpan<byte> buffer) => BitConverter.ToInt16(buffer);
+
+        internal static string ReadString(ArraySegment<byte> buffer, Encoding encoding)
         {
-            if (size < 0 || size > _buffer.Count - _pos)
+            int size = ReadSize(buffer, encoding);
+            if (size == 0)
             {
-                throw new IndexOutOfRangeException($"cannot skip {size} bytes");
+                return "";
+            }
+            else
+            {
+                return _utf8.GetString(buffer.AsSpan(size < 254 ? 1 : 4, size));
```

Comment:
I think there is a bug here from my previous InputStream PR that introduced this static method

```suggestion
                return _utf8.GetString(buffer.AsSpan(size < 255 ? 1 : 5, size));
```

Reply:
Fixed.